tests/integrations/realcluster/cluster.go

```diff
@@ -17,173 +17,245 @@ package realcluster
-			--kv.binpath ./third_bin/tikv-server \
-			--db.binpath ./third_bin/tidb-server \ 
-			--tiflash.binpath ./third_bin/tiflash \
-			--tso.binpath ./bin/pd-server \
-			--scheduling.binpath ./bin/pd-server \
-			--pd.config ./tests/integrations/realcluster/pd.toml \
-			> `+filepath.Join(playgroundLogDir, tag+".log")+` 2>&1 & `)
-		} else {
-			runCommand("sh", "-c",
-				tiupBin+` playground nightly --kv 3 --tiflash 1 --db 1 --pd 3 \
-			--without-monitor --tag `+tag+` \
-			--pd.binpath ./bin/pd-server \
-			--kv.binpath ./third_bin/tikv-server \
-			--db.binpath ./third_bin/tidb-server \
-			--tiflash.binpath ./third_bin/tiflash \
-			--pd.config ./tests/integrations/realcluster/pd.toml \
-			> `+filepath.Join(playgroundLogDir, tag+".log")+` 2>&1 & `)
+		defer logutil.LogPanic()
+		playgroundOpts := []string{
+			fmt.Sprintf("--kv %d", defaultTiKVCount),
+			fmt.Sprintf("--tiflash %d", defaultTiFlashCount),
+			fmt.Sprintf("--db %d", defaultTiDBCount),
+			fmt.Sprintf("--pd %d", defaultPDCount),
+			"--without-monitor",
+			fmt.Sprintf("--tag %s", c.tag),
+		}
+
+		if c.mode == "ms" {
+			playgroundOpts = append(playgroundOpts,
+				"--pd.mode ms",
+				"--tso 1",
+				"--scheduling 1",
+			)
 		}
+
+		cmd := fmt.Sprintf(`%s playground nightly %s %s > %s 2>&1 & `,
+			tiupBin,
+			strings.Join(playgroundOpts, " "),
+			buildBinPathsOpts(c.mode == "ms"),
+			filepath.Join(playgroundLogDir, c.tag+".log"),
+		)
+		_, err := runCommandWithOutput(cmd)
+		re.NoError(err)
 	}()
 
 	// Avoid to change the dir before execute `tiup playground`.
 	time.Sleep(10 * time.Second)
-	require.NoError(t, os.Chdir(curPath))
+	re.NoError(os.Chdir(curPath))
+}
+
+// collectPids will collect the pids of the processes.
+func (c *cluster) collectPids() error {
+	output, err := runCommandWithOutput(fmt.Sprintf("pgrep -f %s", c.tag))
+	if err != nil {
+		return fmt.Errorf("failed to collect pids: %v", err)
+	}
+
+	for _, pidStr := range strings.Split(strings.TrimSpace(output), "\n") {
+		if pid, err := strconv.Atoi(pidStr); err == nil {
+			c.pids = append(c.pids, pid)
+		}
+	}
+	return nil
 }
 
-func waitTiupReady(t *testing.T, tag string) {
-	const (
-		interval = 5
-		maxTimes = 20
-	)
-	log.Info("start to wait TiUP ready", zap.String("tag", tag))
-	for i := range maxTimes {
-		err := runCommand(tiupBin, "playground", "display", "--tag", tag)
-		if err == nil {
-			log.Info("TiUP is ready", zap.String("tag", tag))
-			return
+func (c *cluster) waitReady() {
+	re := c.re
+	log.Info("start to wait TiUP ready", zap.String("tag", c.tag))
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-time.After(100 * time.Second):
+			re.FailNowf("TiUP is not ready after timeout, tag: %s", c.tag)
+		case <-ticker.C:
+			log.Info("check TiUP ready", zap.String("tag", c.tag))
+			cmd := fmt.Sprintf(`%s playground display --tag %s`, tiupBin, c.tag)
+			output, err := runCommandWithOutput(cmd)
+			if err == nil {
```

Comment:
Don't we need to check its output also?

Reply:
I may change the logic here in the next PR.